src/lib.rs

```diff
@@ -1,40 +1,97 @@
+            .ok_or_else(|| {
+                format!(
+                    "Cannot represent {} as {}",
+                    type_name::<$actual>(),
+                    type_name::<$expected>()
+                )
+            })
+            .unwrap()
+    };
+}
+
+/// The basic GeoTIFF struct. This includes any metadata as well as the actual image data.
 ///
-/// It is primarily used within a routing application that needs to parse digital elevation models.
-/// As such, other use cases are NOT tested (for now).
-impl TIFF {
-    /// Opens a `.tiff` file at the location indicated by `filename`.
-    pub fn open(filename: &str) -> Result<Box<TIFF>> {
-        let tiff_reader = TIFFReader;
-        tiff_reader.load(filename)
-    }
+/// The image data has a size of width * length * num_samples
+#[derive(Debug)]
+pub struct GeoTIFF {
+    pub dimensions: (usize, usize),
+    pub num_samples: usize,
+    image_data: ImageData,
+}
+
+impl GeoTIFF {
+    pub fn read<R: Read + Seek>(reader: R) -> TiffResult<Self> {
+        let mut decoder = Decoder::new(reader)?;
 
-    /// Gets the value at a given coordinate (in pixels).
-    pub fn get_value_at(&self, lon: usize, lat: usize) -> usize {
-        self.image_data[lon][lat][0]
+        let dimensions = decoder
+            .dimensions()
+            .map(|(width, height)| (width as usize, height as usize))?;
+        let num_samples = match decoder.find_tag(Tag::SamplesPerPixel)? {
+            None => 1,
+            Some(value) => value.into_u16()? as usize,
+        };
+        let image_data = match decoder.read_image()? {
+            DecodingResult::U8(data) => ImageData::U8(data),
+            DecodingResult::U16(data) => ImageData::U16(data),
+            DecodingResult::U32(data) => ImageData::U32(data),
+            DecodingResult::U64(data) => ImageData::U64(data),
+            DecodingResult::F32(data) => ImageData::F32(data),
+            DecodingResult::F64(data) => ImageData::F64(data),
+            DecodingResult::I8(data) => ImageData::I8(data),
+            DecodingResult::I16(data) => ImageData::I16(data),
+            DecodingResult::I32(data) => ImageData::I32(data),
+            DecodingResult::I64(data) => ImageData::I64(data),
+        };
+
+        Ok(Self {
+            dimensions,
+            num_samples,
+            image_data,
+        })
     }
-}
 
-/// Overwrite default display function.
-impl fmt::Display for TIFF {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "TIFF(Image size: [{}, {}, {}], Tag data: {:?})",
-               self.image_data.len(), self.image_data[0].len(),
-               self.image_data[0][0].len(), self.ifds)
+    pub fn get_value_at<T: FromPrimitive + 'static>(&self, x: usize, y: usize, sample: usize) -> T {
+        let GeoTIFF {
+            dimensions: (image_width, ..),
+            num_samples,
+            image_data,
+            ..
+        } = self;
+
+        if sample >= *num_samples {
```

Comment:
Could also match by turning `sample` into a reference instead of dereferencing `num_samples`?

```suggestion
        if &sample >= num_samples {
```

Reply:
Will do!